src/net/ossl.cc

```diff
@@ -816,8 +854,30 @@ class session : public enable_shared_from_this<session>, public session_impl {
 
 
 private:
+    std::shared_ptr<X509> get_peer_certificate() const {
+        auto res = SSL_get_verify_result(_ssl.get());
+        auto peer_cert = make_shared_x509_cert(SSL_get1_peer_certificate(_ssl.get()));
```

Comment:
Without my suggestion, wouldn't `X509_up_ref` be required here? Or is that intentional, I only read the docs, not the source code, so I have no idea how ownership works.